### PR TITLE
issue 235: Compiler should check qooxdoo-range and warn/abort if not matched

### PR DESCRIPTION
**Pull request by hkollmann**
Second try after https://github.com/qooxdoo/qooxdoo-compiler/pull/268.
New PR to cleanup commits and integrate changes of @cboulanger in https://github.com/qooxdoo/qooxdoo-compiler/pull/271

This PR adds following features:
new section in Manifest,json:
 "requires": {
      "qooxdoo-sdk": "^6.0.0-alpha",
      "qooxdoo-compiler": "^0.2.40",
       "qooxdoo/qxl.versionlabel": "^10.1.1",
       "qooxdoo/qxl.formdemo": "^0.1.1"
   }
}
Compiler do a semver check of required contribs. This check is be done for all installed contribs.
Error is thrown if version do not match. Should this be overrun by parameter?
Compiler do a semver check for sdk and compiler version. Error is thrown if version do not match, Should this be overrun by parameter?
install.js adds all required contribs to contrib.json. This is be done recursive.
install.js adds intalled contrib to requires section of the current project. Should this be disabled by parameter?
"qooxdoo-range" is be marked as deprecated. Instead of info.qooxdoo-range requires.qooxdoo-sdk should be used.
Changed documentation of Manifest,json to reflect this.
implemented issue 244: allow qx install library@version
implemented issue 196: qx contrib install: more helpful message if no compatible release #19

### CODE REVIEW

File: docs/cli/contrib.md [hunk 1]
```diff
@@ -238,25 +238,25 @@ developed with a particular qooxdoo version).
 The qooxdoo framework version can be found in the [top level `version.txt`
 file](https://github.com/qooxdoo/qooxdoo/blob/master/version.txt). The contrib
 declares its compatibility with qooxdoo versions using the `qooxdoo-versions` OR
-`qooxdoo-range` entries in `Manifest.json` (See [this
-example](https://github.com/cboulanger/qx-contrib-Dialog/blob/master/Manifest.json#L21)).
+`requires.libraries.qx` entries in `Manifest.json` (See [this
```
Comment (cboulanger) [inline docs/cli/contrib.md]: Same here.

File: docs/cli/contrib.md [hunk 2]
```diff
@@ -238,25 +238,25 @@ developed with a particular qooxdoo version).
...
+example](https://github.com/qooxdoo/qxl.widgetbrowser/blob/master/Manifest.json#L47)).
 `qooxdoo-versions` takes an array of version numbers. You need to specify each
 and every version that you want to support, and any new qooxdoo version will
 break compatibility. This is part of the original, now defunct contrib system.
 It will be supported for some time, but might be deprecated in the future.
 
-We strongly suggest to use the `qooxdoo-range` key instead, which takes a
+We strongly suggest to use the `requires.libraries.qx` key instead, which takes a
```
Comment (cboulanger) [inline docs/cli/contrib.md]: and here.

File: docs/cli/contrib.md [hunk 3]
```diff
@@ -238,25 +238,25 @@ developed with a particular qooxdoo version).
...
 [semver range string](https://github.com/npm/node-semver#ranges). This allows
 for a much more flexible and automated dependency management. You can, for
 example, declare that the contrib will be compatible with qooxdoo versions
 starting with 5.02 up until version 6, i.e. as long as there is no breaking
 change (which is guaranteed by the semver specs), using `5.0.2 - 6.x` as the
-`qooxdoo-range`. The `qx contrib` commands handle the compatibility data
+`requires.libraries.qx`. The `qx contrib` commands handle the compatibility data
```
Comment (cboulanger) [inline docs/cli/contrib.md]: and here.

File: docs/cli/contrib.md [hunk 4]
```diff
@@ -238,25 +238,25 @@ developed with a particular qooxdoo version).
...
 generated by semver strictly, and the compiler will enforce these compatibility
-restraints (TBD).
+restraints.
 
 In future versions, contribs can also declare their *dependencies* on other
```
Comment (cboulanger) [inline docs/cli/contrib.md]: Isn't this what is implemented by your PR? If so, it's not in a future version.

File: lib/qx/tool/cli/commands/contrib/Install.js [hunk 5]
```diff
@@ -69,165 +73,221 @@ qx.Class.define("qx.tool.cli.commands.contrib.Install", {
...
+        await this.installApplication(current_download_path);
+        await this.installDependencies(current_download_path);
         if (!this.argv.quiet){
-          console.info(`>>> Installed ${info.name} (${argv.contrib_uri}, ${info.version})`);
+          console.info(`>>> Installed ${info.name} (${current_uri}, ${info.version})`);
         }
       }
       if (!found) {
-        throw new qx.tool.cli.Utils.UserError(`The contrib library identified by '${argv.contrib_uri}' could not be found.`);
+        throw new qx.tool.cli.Utils.UserError(`The contrib library identified by '${repoName}' could not be found.`);
```
[the remark this answers is not in the document]
Comment (hkollmann) [reply lib/qx/tool/cli/commands/contrib/Install.js]: The parameter is needed here! I renamed it to uriName

File: lib/qx/tool/cli/commands/contrib/Install.js [hunk 6]
```diff
@@ -69,165 +73,221 @@ qx.Class.define("qx.tool.cli.commands.contrib.Install", {
...
 
-        if (contrib_path) {
-          // if a path component exists, only install the contrib in this path
-          if (path.dirname(filename) !== contrib_path) continue;
-          current_download_path = path.join(download_path, contrib_path);
-        } else if (path.dirname(filename) !== ".") {
-          // else, skip if no manifest exists in the root of the repo
-          continue;
-        }
+        if (contrib_path && (path.dirname(filename) !== contrib_path)) {
```
[the remark this answers is not in the document]
Comment (hkollmann) [reply lib/qx/tool/cli/commands/contrib/Install.js]: What's the use case?

[the remark this answers is not in the document]
Comment (hkollmann) [reply lib/qx/tool/cli/commands/contrib/Install.js]: With your version i can not install the ergobyte/qookery as example in a whole.

[the remark this answers is not in the document]
Comment (cboulanger) [reply lib/qx/tool/cli/commands/contrib/Install.js]: Does that work? When I tried that, it threw an error,  that’s why I disallowed installation of the repo when there is no Manifest in its root dir. If it does work, your change is fine. Can you add a test for it?

[the remark this answers is not in the document]
Comment (hkollmann) [reply lib/qx/tool/cli/commands/contrib/Install.js]: Yes, it do. Testcase added and fix if contrib is partly removed

**Pull request merged**